File: python/115. 不同的子序列.py
class Solution_115:
    def numDistinct(self,s,p):
        m, n = len(s), len(t)
        if (m == 0 or n == 0): return 0
        dp = [0] * (n + 1)
        dp[0] = 1
        for i in range(1, m + 1):
            for j in range(n, 0, -1):
                if (s[i - 1] == t[j - 1]):
                    dp[j] += dp[j - 1]
        return dp[n]
if __name__ == '__main__':
    s=input()
    t=input()
    s_115=Solution_115()
    print(s_115.numDistinct(s,t))

File: c++/115. 不同的子序列.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;
class Solution_115 {
public:
    int numDistinct(string s, string t) {
        int m = s.size(), n = t.size();
        if (m == 0 || n == 0) return 0;
        //  vector<vector<int>> dp(m+1,vector<int>(n+1));
        //  for(int i=0;i<=m;++i) {
        //      dp[i][0]=1;
        //  }
        //  for(int i=1;i<=m;++i){
        //      for(int j=1;j<=n;++j){
        //          if(INT_MAX-dp[i-1][j]>=dp[i-1][j-1]&&s[i-1]==t[j-1]){
        //              dp[i][j]=dp[i-1][j-1]+dp[i-1][j];
        //          }
        //          else{
        //              dp[i][j]=dp[i-1][j];
        //          }
        //      }
        //  }
        //  return dp[m][n];
        vector<int> dp(n + 1, 0);
        dp[0] = 1;
        for (int i = 1; i <= m; ++i) {
            for (int j = n; j >= 1; --j) {
                if (INT_MAX - dp[j] >= dp[j - 1] && s[i - 1] == t[j - 1]) {
                    dp[j] += dp[j - 1];
                }
            }
        }
        return dp[n];
    }
};
void main_115() {
    string s, t;
    cin >> s;
    cin >> t;
    Solution_115 s_115;
    cout<<s_115.numDistinct(s, t)<<endl;
}